### PR TITLE
[Move] [break] Support TypeTag and StructTag to canonical string with prefix

### LINKED ISSUE #2395: [Move] Add prefix to to_canonical_string
The TypeTag::to_canonical_string returns a string like:

`00000000000000000000000000000001::string::String` or
  `0000000000000000000000000000000a::module_name1::type_name1<0000000000000000000000000000000a::module_name2::type_name2<u64>>`

But we can not parse the string via `TypeTag::from_str`, and it will cause some issues like #2394 

Maybe we should add the `0x` prefix?

https://github.com/MystenLabs/sui/blob/58fd1d1d2cadec6fb2cc6b85df61727db56a2ad2/external-crates/move/crates/move-core-types/src/language_storage.rs#L62-L109

This will be a break change.

### PR DESCRIPTION
**Pull request by baichuan3**
## Summary

1. Support `TypeTag`, `StructTag` and `AccountAddress` to canonical string with prefix `0x`. 
2. Unify the canonical and non-canonical string format to
`0x00000000000000000000000000000001::string::String` or
`0x0000000000000000000000000000000a::module_name1::type_name1<0x0000000000000000000000000000000a::module_name2::type_name2<u64>>`

Other impacts include possible break of SDK and front-end applications, please confirm @wow-sven 

1. The `to_canonical_string()` method of TypeTag, StructTag and AccountAddress returns a prefix containing 0x by default
1. The generated type name contains the 0x prefix
3. Generation of NamedObjectID
4. Generation of some field keys

- Closes #2395

### CODE REVIEW

File: moveos/moveos-types/src/move_types.rs [hunk 1]
```diff
@@ -369,7 +374,11 @@ pub fn parse_type_tag(s: &str) -> Result<TypeTag> {
 pub fn parse_struct_tag(s: &str) -> Result<StructTag> {
     let s = s.trim();
     if s.starts_with("0x") {
-        StructTag::from_str(s)
+        if let Ok(struct_tag) = StructTag::from_str(s) {
+            Ok(struct_tag)
+        } else {
+            struct_tag_from_canonical_string(s)
```
Comment (jolestar) [inline moveos/moveos-types/src/move_types.rs:380]: We can directly call `StructTag::from_str` now and remove the struct_tag_from_canonical_string.

File: moveos/moveos-types/src/move_types.rs [hunk 2]
```diff
@@ -383,13 +392,22 @@ mod tests {
...
+            "0x0000000000000000000000000000000000000000000000000000000000000005::test::struct";
         let test_type_tag = TypeTag::from_str(type_tag_str)?;
         assert_eq!(type_tag_str, test_type_tag.to_string());
         assert_eq!(type_tag_canonical_str, test_type_tag.to_canonical_string());
         Ok(())
     }
 
+    #[test]
+    fn test_strange_format_type_tag() -> Result<()> {
+        let type_tag_str = "0x900f053234b0ba66ad062b277896b28e049f2813d388da375efcd54b6e429dbe::coin::Coin<0000000000000000000000000000000000000000000000000000000000000003::gas_coin::GasCoin>";
```
Comment (jolestar) [inline moveos/moveos-types/src/move_types.rs:404]: This is an invalid type_tag_str, we do not need to support it.

**Review by jolestar — COMMENTED** [review]
The named object ID in the unit test and the testsuit *.feature files also need to be updated.

File: moveos/moveos-types/src/move_types.rs (hunk 1 above)
Comment (baichuan3) [reply moveos/moveos-types/src/move_types.rs]: done

**Review by popcnt1 — APPROVED** [review]

**Review by jolestar — APPROVED** [review]

**Pull request merged**